HPCC-16755 Release log queue lock after a queue item is read

rpastrana opened the pull request:
In the existing code, after a log thread gets log queue lock, it
does not release the lock before finishing all of queued log items.
ESP logging manager may wait for a long time to add a log request
into the queue if the log thread is working on the queued items.
In this fix, the log thread releases the queue lock as soon as it
reads a queue item from the queue.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

rpastrana commented:
@gaojx please review

HPCCSmoketest commented:
Automated Smoketest
Sha: 04a8985df5c640ba786259c03d6ad681afb46831
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 90.4 seconds |

Install hpccsystems-platform-community_6.2.0-rc6.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|88 |88 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented on esp/logging/logginglib/logthread.cpp:
```diff
@@ -414,3 +400,24 @@ IEspUpdateLogRequestWrap* CLogThread::unserializeLogRequestContent(const char* l
     return new CUpdateLogRequestWrap(guid, opt, buffer.str());
 };
+
+IEspUpdateLogRequestWrap* CLogThread::readWriteJobQueue(IEspUpdateLogRequestWrap* jobToWrite)
```
I think extracting some of the code into a separate function is a good idea, but combining two different actions into the same function is not.  The only thing they share in common is that they need a critical section - they don't even share a parameter or a return value.  Much better would be "void writeJobQueue(item)" and IRequest * readJobQueue()

In this case a minimal change would probably be better, witht eh refactor as a separate commit.  I think it request the critical section to be moved into a local scope, something like:

```
   {
       CriticalBlock b(logQueueCrit);
       logRequest  = (IEspUpdateLogRequestWrap*)logQueue.dequeue();
   }

```

richardkchapman commented:
Commit message still refers to HPCC-9999

rpastrana committed:
HPCC-16755 Release log queue lock after a queue item is read

- Decreases the amount of time spent within critset
- provides methods to read/write to/from log queue

Signed-off-by: wangkx <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@ghalliday please review (squashed commits in order to fix Kevin's commit message)

ghalliday commented on esp/logging/logginglib/logthread.cpp:
```diff
@@ -414,3 +400,29 @@ IEspUpdateLogRequestWrap* CLogThread::unserializeLogRequestContent(const char* l
+IEspUpdateLogRequestWrap* CLogThread::readJobQueue()
+{
+	CriticalBlock b(logQueueCrit);
+	if (logQueue.ordinality() > 0)
```
minor: This test is redundant since dequeue() already returns null if it is empty.

ghalliday commented on esp/logging/logginglib/logthread.cpp:
```diff
@@ -414,3 +400,29 @@ IEspUpdateLogRequestWrap* CLogThread::unserializeLogRequestContent(const char* l
+
+void CLogThread::writeJobQueue(IEspUpdateLogRequestWrap* jobToWrite)
+{
+	if (jobToWrite)
```
whitespace: This has been indented with tabs rather than space characters.

ghalliday commented on esp/logging/logginglib/logthread.cpp:
```diff
@@ -414,3 +400,29 @@ IEspUpdateLogRequestWrap* CLogThread::unserializeLogRequestContent(const char* l
+
+IEspUpdateLogRequestWrap* CLogThread::readJobQueue()
+{
+	CriticalBlock b(logQueueCrit);
```
whitespace: tabs used instead of spaces.

HPCCSmoketest commented:
Automated Smoketest
Sha: 2119f5308f1435c9a0d617ac4900e6118bb7a7a9
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.675 seconds |

Install hpccsystems-platform-community_6.2.0-rc6.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|88 |88 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

rpastrana committed:
HPCC-16755 Code review changes

- replaces tabs with spaces
- removes superfluous conditional

Signed-off-by: rpastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@ghalliday one last review. Thanks.
